hypothesis-python/tests/watchdog/test_database.py

```diff
@@ -0,0 +1,152 @@
+)
+
+from tests.cover.test_database_backend import _database_conforms_to_listener_api
+
+# we need real time here, not monkeypatched for CI
+time_sleep = time.sleep
+
+
+def test_database_listener_directory():
+    # this test is very expensive because we wait between every rule for the
+    # filesystem observer to fire. Limit examples/step count as much as possible.
+    _database_conforms_to_listener_api(
+        lambda path: DirectoryBasedExampleDatabase(path),
+        flush=lambda _db: time_sleep(0.2),
+        supports_value_delete=False,
+        # expensive flush makes shrinking take forever
+        parent_settings=settings(
+            max_examples=5, stateful_step_count=10, phases=set(Phase) - {Phase.shrink}
+        ),
+    )
+
+
+def test_database_listener_multiplexed(tmp_path):
+    db = MultiplexedDatabase(
+        InMemoryExampleDatabase(), DirectoryBasedExampleDatabase(tmp_path)
+    )
+    events = []
+
+    def listener(event):
+        events.append(event)
+
+    db.add_listener(listener)
+
+    db.save(b"a", b"a")
+    time_sleep(0.2)
+    assert events == [("save", (b"a", b"a"))] * 2
+
+    db.remove_listener(listener)
+    db.delete(b"a", b"a")
+    db.save(b"a", b"b")
+    time_sleep(0.2)
+    assert events == [("save", (b"a", b"a"))] * 2
+
+    db.add_listener(listener)
+    db.delete(b"a", b"b")
+    db.save(b"a", b"c")
+    time_sleep(0.2)
+    # InMemory database fires immediately, while DirectoryBased has to
+    # wait for filesystem listeners. Therefore the events can arrive out of
+    # order. Test a weaker set property, disregarding ordering.
+    assert set(events[2:]) == {
+        # InMemory
+        ("delete", (b"a", b"b")),
+        ("save", (b"a", b"c")),
+        # DirectoryBased
+        ("delete", (b"a", None)),
+        ("save", (b"a", b"c")),
+    }
+
+
+def test_database_listener_directory_explicit(tmp_path):
+    db = DirectoryBasedExampleDatabase(tmp_path)
+    events = []
+
+    def listener(event):
+        events.append(event)
+
+    db.add_listener(listener)
+
+    db.save(b"k1", b"v1")
+    time_sleep(0.2)
+    assert events == [("save", (b"k1", b"v1"))]
+
+    db.remove_listener(listener)
+    db.delete(b"k1", b"v1")
+    db.save(b"k1", b"v2")
+    time_sleep(0.2)
+    assert events == [("save", (b"k1", b"v1"))]
+
+    db.add_listener(listener)
+    db.delete(b"k1", b"v2")
+    db.save(b"k1", b"v3")
+    time_sleep(0.2)
+    assert events[1:] == [
+        ("delete", (b"k1", None)),
+        ("save", (b"k1", b"v3")),
+    ]
+
+    # moving into a nonexistent key
+    db.move(b"k1", b"k2", b"v3")
+    time_sleep(0.2)
+    # moving back into an existing key
+    db.move(b"k2", b"k1", b"v3")
+    time_sleep(0.2)
+
+    if sys.platform.startswith("darwin"):
+        expected = [
+            ("delete", (b"k1", b"v3")),
+            ("save", (b"k2", b"v3")),
+            ("delete", (b"k2", b"v3")),
+            ("save", (b"k1", b"v3")),
+        ]
+    elif sys.platform.startswith("win"):
```

Comment:
the linux details for `move` here are weird, but not really harmful, just redundant/inefficient. I suspect this is at the os/watchdog level but don't have a linux machine to test further. We use `os.renames` for `db.move` with a fallback to save/delete, which may explain some of the divergent behavior.

Reply:
github "codespaces" might be helpful to debug?